### ISSUE TITLE
Indent rule bug in arrays when first value has indent

### ISSUE THREAD

**Issue by malyw** [eslint/eslint#4161]
**ESlint config:**
```
"indent": [
            2,
            4
        ],
```

**Current behavior:**

*Code 1:*
```
var a = [
                'a',
                'b'
        ];
```
**Error: "Expected indentation of 12 space characters but found 16 (indent)."**

*Code 2:*
```
var a = ['a',
                'b'
        ];
```
**Error: None**

**Expected behavior:**
Both options don't throw any error as for JSHint option `"indent": 4`
as described on http://eslint.org/docs/rules/indent.html#compatibility

**Other**
ESlint version: 1.6.0

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
When we say compatibility we mean that we have a similar rule but that doesnt necessarily mean it functions exactly like that.

**Code **
```js
/* eslint indent: [2, 4] */
var a = [
                'a',
                'b'
        ];
```
```sh
3:17 - Expected indentation of 4 space characters but found 16. (indent)
4:17 - Expected indentation of 4 space characters but found 16. (indent)
5:9 - Expected indentation of 0 space characters but found 8. (indent)
```

Expectations: Correct behavior

**Code 2**
```js
/* eslint indent: [2, 4] */
var a = ['a',
                'b'
        ];
```
```sh
No errors
```

Expectations: Incorrect behavior so its a bug we need to fix this.
Expected errors:
```sh
3:17 - Expected indentation of 4 space characters but found 16. (indent)
4:9 - Expected indentation of 0 space characters but found 8. (indent)
```

This is how I see this going.

**Comment by jFransham** [comment]
I don't know if this is the same issue, but I have a problem with arrow function syntax and first-element-indented arrays where the indent checker thinks it can never be indented enough.

```javascript
if (
	[
		'one',
		'two',
		'three',
	].find((el) =>
		[                  // <| error: Expected indentation of 3 tab characters but found 2. (indent)
			'one',         // <| error: Expected indentation of 4 tab characters but found 3. (indent)
			'two',         // <| error: Expected indentation of 4 tab characters but found 3. (indent)
			'three',       // <| error: Expected indentation of 4 tab characters but found 3. (indent)
		].indexOf(el) >= 0 // <| error: Expected indentation of 3 tab characters but found 2. (indent)
	)
) {
	console.log('Testing');
}

if (
	[
		'one',
		'two',
		'three',
	].find((el) =>
			[                  // <| error: Expected indentation of 4 tab characters but found 3. (indent)
				'one',         // <| error: Expected indentation of 5 tab characters but found 4. (indent)
				'two',         // <| error: Expected indentation of 5 tab characters but found 4. (indent)
				'three',       // <| error: Expected indentation of 5 tab characters but found 4. (indent)
			].indexOf(el) >= 0 // <| error: Expected indentation of 4 tab characters but found 3. (indent)
	)
) {
	console.log('Testing');
}
```

**Comment by MoOx** [comment]
Getting a similar issue with the following code

```js
const a = ExtractTextPlugin.extract(
  "style-loader",
  [
    "css-loader",
    "postcss-loader",
  ].join("!"),
)
```


```
  10:3   error  Expected indentation of 0 space characters but found 2    indent
  11:5   error  Expected indentation of 2 space characters but found 4    indent
  12:5   error  Expected indentation of 2 space characters but found 4    indent
  13:3   error  Expected indentation of 0 space characters but found 2    indent
```

It looks like bug.

Note that the following code (the same without the `const a =`) does not trigger the issue

```
ExtractTextPlugin.extract(
  "style-loader",
  [
    "css-loader",
    "postcss-loader",
  ].join("!"),
)
```

This seems inconsistent.

**Comment by sstern6** [comment]
@gyandeeps has there been any update on this?

Would like to take a swing at it, this will be my second issue patch in ESLint so will have some questions if I can take this one on?

Let me know

Thank you

**Comment by ilyavolodin** [comment]
@sstern6 Feel free to take over this. This issue has been sitting for a while without updates, so I'm pretty sure nobody is working on it currently! Let us know if you have any questions/need help!

**Comment by gyandeeps** [comment]
@sstern6 Soory for late response but exactly what @ilyavolodin said.

**Comment by sstern6** [comment]
@ilyavolodin @gyandeeps thanks will probably have this up by mid to end of this coming week.  Will definitely have questions for you!

Thank you

**Comment by sstern6** [comment]
@ilyavolodin think I have found where the issue is coming from but I am having trouble identifying what file the fix should come from.

When you run the code:

```
var a = ['a',
                'b'
        ];
```
`ArrayExpression(node)` is never called, meaning this format of declaring a var never gets assigned to an ArrayExpression.  

However, when the code is:

```
var a = [
                'a',
                'b'
        ];
```
`ArrayExpression(node)` IS called triggering `checkIndentInArrayOrObjectBlock(node)`.

Does it seem like im on the right track?

If so, what file is ArrayExpression being set, because if I am on the right track, this doesnt seem like it can be fixed in `lib/rules/index.js`.

Thank you

**Comment by vitorbal** [comment]
If you drop the example code in astexplorer, you can see the ArrayExpression node in there: https://astexplorer.net/#/mdfJnwsl7S
So I don't think that's the problem. Are you sure `checkIndentInArrayOrObjectBlock(node)` is not being called? Maybe it is being called, but going into one of the early `return` statements of that function.

**Comment by sstern6** [comment]
@vitorbal I put a `console.log` inside of this function https://github.com/eslint/eslint/blob/master/lib/rules/indent.js#L891 right before `checkIndentInArrayOrObjectBlock(node)`. When analyzing the code:

```
/*eslint indent: [2,4]*/
var a = ['a',
                'b'
        ];
```

The console.log is never reached.  However, when running with:

```
/*eslint indent: [2,4]*/
var a = [
                'a',
                'b'
        ];
```

The console.log is reached.

That is how I came to the conclusion.  Does that sound reasonable?  Will continue to dig into this.

**Comment by sstern6** [comment]
@vitorbal ya through some additional searching it wasnt  what I thought, you were right, `checkIndentInArrayOrObjectBlock` was exiting too early.

Will have a fix up hopefully in the next day!  Thanks for the advice

**Comment by sstern6** [comment]
@ilyavolodin @vitorbal @gyandeeps found where the issue is coming from.  Looks like https://github.com/eslint/eslint/blob/master/lib/rules/indent.js#L627.  What Im curious about is the best way to handle this.  Given the example above, I found if you check:

```
if (elements.length > 0 && elements[0].loc.start.line === node.loc.start.line) {
    if (elements[0].type !== "Literal") {  // my added logic
        return;
    }
}
```

it handles the example @gyandeeps and @malyw showed thats breaking and doesnt break any current tests.  However, I was thinking of other cases this could happen, ie:

```
/*eslint indent: [2, 4]*/
var x = [{a: 1},
            {b: 2}
        ];
```

Errors: None; (Should have 2 errors)
or

```
/*eslint indent: [2, 4]*/
var x = [
          {a: 1},
            {b: 2}
        ];
```
Errors:  (Should have 3 errors)
`error  Expected indentation of 0 space characters but found 8`

How in depth should we get for edge cases?  Is checking for the 'Literal' enough?  

Finally, something else I noticed was that if you have an array of integers it works as expected in the given test format we have been using. So it seems dependent on the type of value that is in the ArrayExpression.

```
var x = [1,
        2];

var x = [
        1,
        2];
```

I would appreciate a point in the right direction.

Thanks

**Comment by gyandeeps** [comment]
@sstern6 Thanks for all the efforts. I wont be able to reply for few days as i workin on some critical issue at work. Just wanted to let u know.
Posting this message without reading the whole thread. Sorry buddy but I will be back (Thanks for understanding). Again I want to thank u for all ur efforts.

**Comment by sstern6** [comment]
Will have a PR open for this shortly.
